tools/applink/src/devConsole/pipeline/buildScheduler.ts

```diff
@@ -0,0 +1,289 @@
+
+        allPackagesToRebuild.forEach((pkg) => (this.packageStates[pkg].dirty = true));
+
+        for (const packageToRebuild of allPackagesToRebuild) {
+            const packageState = this.packageStates[packageToRebuild];
+
+            if (packageState.buildRunning) {
+                packageState.buildDirty = true;
+            }
+        }
+
+        this.eventBus.post(buildScheduled(changes, dependenciesToRebuild));
+    };
+
+    /*
+     * Do bookkeeping after the build finishes:
+     *
+     * -  On success, clean failed indicator; if the build was not dirty, the package is clean
+     * -  On fail, turn on failed indicator; if the build was not dirty, the package is clean. otherwise it
+     *    stays dirty.
+     *
+     * Note that packages that stay dirty because of dirty build will be rebuild on next triggerBuilds().
+     */
+    private processBuildFinished = (event: BuildFinished): void => {
+        const { exitCode, packageName } = event.body;
+        const packageState = this.packageStates[packageName];
+
+        if (exitCode) {
+            packageState.dirty = packageState.buildDirty;
+            packageState.failed = true;
+
+            appLogError(`Build of ${packageName} has failed.`);
+        } else {
+            packageState.dirty = packageState.buildDirty;
+            packageState.failed = false;
+
+            if (!packageState.dirty) {
+                this.buildsToGetClean.add(packageName);
+            } else {
+                appLogWarn(
+                    `Build of ${packageName} has succeeded however more changes to this package were register in the meanwhile. Have to rebuild.`,
+                );
+            }
+        }
+
+        packageState.buildDirty = false;
+        packageState.buildRunning = false;
+    };
+
+    private triggerBuilds = (): void => {
+        this.findDirtyLeaves().forEach((pkg) => {
+            const packageState = this.packageStates[pkg];
+
+            packageState.buildRequested = true;
+            packageState.buildRunning = false;
+            packageState.buildDirty = false;
+
+            this.eventBus.post(buildRequested(pkg));
+        });
+    };
+
+    /*
+     * Find dirty leaves - those are packages which themselves are dirty but only depend
+     * on packages that are clean and successfully built.
+     *
+     * This may find multiple packages = those can be naturally built in parallel.
+     */
+    private findDirtyLeaves = (): string[] => {
+        const dirtyPackages: string[] = Object.entries(this.packageStates)
+            .filter(([, state]) => state.dirty && !state.buildRequested && !state.buildRunning)
+            .map(([packageName]) => packageName);
+
+        if (!dirtyPackages.length) {
+            return [];
+        }
+
+        const packagesWithCleanDeps = dirtyPackages.filter((pkg) => {
+            return Object.values(this.dependencyGraph!.outgoing[pkg] ?? []).every((dependency) => {
```

Comment:
i would keep the '!' so that the code bombs if it is undefined. it would indicate that something stateful is wrong and needs to be handled explicitly, probably somewhere else (the call to find dirty leaves should not happen if there is no dependency graph available)

Reply:
Fair enough